python/ray/serve/_private/http_proxy.py

```diff
@@ -75,9 +78,32 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
             assignment_task.cancel()
         try:
             object_ref = await assignment_task
-            result = await object_ref
-            client_disconnection_task.cancel()
-            break
+
+            # NOTE (shrekris-anyscale): when the gcs, Serve controller, and
+            # some replicas crash simultaneously (e.g. if the head node crashes),
+            # requests to the dead replicas hang until the gcs recovers.
+            # This asyncio.wait can kill those hanging requests and retry them
+            # at another replica. Release tests should kill the head node and
+            # check if latency drops significantly. See
+            # https://github.com/ray-project/ray/pull/29534 for more info.
+
+            _, request_timed_out = await asyncio.wait(
+                [object_ref], timeout=SERVE_REQUEST_PROCESSING_TIMEOUT_S
+            )
+            if request_timed_out:
+                logger.info(
+                    "Request didn't finish within "
+                    f"{SERVE_REQUEST_PROCESSING_TIMEOUT_S} seconds. Retrying "
+                    "with another replica. You can modify this timeout by "
+                    'setting the "SERVE_REQUEST_PROCESSING_TIMEOUT_S" env var.'
+                )
+                client_disconnection_task.cancel()
+                backoff_time_s *= 1.5
```

Comment:
There are some duplicate codes here and `RayActorError`, we can raise RayServeException here and do all the backoff & sleep logics inside the exception?

RayServeException and RayActorError can use same code.

Reply:
Good idea, I moved all the backoff logic to a single conditional.